Description: Change "executeonly" to throw typecheck on gstatetype and devicetype objects
 PS doesn't really have a "devicetype",
 but it is not listed as an acceptable argument type to "executeonly",
 so we probably should not accept it.
 .
 "gstate executeonly" throws typecheck in Acrobat,
 so we should do the same.
 .
 Tested the other "access related" functions
 (noaccess, readonly, rcheck, wcheck, xcheck)
 and all of these accept "gstate" as an argument,
 even though the PLRM says they should not.
 Also our initialization code does things with wcheck and readonly
 on devices and gstates.
 .
 So, this fix is only for "executeonly".
Origin: upstream, http://git.ghostscript.com/?p=ghostpdl.git;h=c1657e3
Author: Nancy Durgin <nancy.durgin@artifex.com>
Forwarded: yes
Last-Update: 2018-10-18
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/psi/ztype.c
+++ b/psi/ztype.c
@@ -174,9 +174,10 @@
 zexecuteonly(i_ctx_t *i_ctx_p)
 {
     os_ptr op = osp;
+    ref_type rtype = r_type(op);
 
     check_op(1);
-    if (r_has_type(op, t_dictionary))
+    if (rtype == t_dictionary || rtype == t_astruct || rtype == t_device)
         return_error(gs_error_typecheck);
     return access_check(i_ctx_p, a_execute, true);
 }
